### ISSUE TITLE
FOQUS Cloud Race Condition:  Results Generator Missing "output"

### ISSUE THREAD

**Issue by boverhof** [CCSI-Toolset/FOQUS#323]
## Description
The "output" of a job description isn't present in the results generator.  The "output" was found to be present both in DynamoDB ( SNS ) and in the S3 file for that job id "" ( output.json ).

FOQUS bombs when it hits this job result because it is missing the output field.

## Diagnosis

SNS update messages are being sent to a lambda function "foqus-sns-update".  These messages are sent in order but may be processes in parallel or out-of-order.  

The SNS update "( event=status, status=success )" is being processed "0.1" seconds before the "( event=output )"

If a results generator page requests happens in this interval of "0.1" seconds it will retrieve a job description without the "output".

## CloudWatch Logs for lambda foqus-sns-update
```
2018-11-15T18:07:26.308Z	4ee1cc74-e901-11e8-94c3-89b452e500e0	process_job_event:
{
    "status": "success",
    "resource": "job",
    "rc": 0,
    "instanceid": "i-065181f20cd90d927",
    "consumer": "5bb40eeb-1ab6-4918-bd2c-4c5e7cf6b67d",
    "event": "status",
    "jobid": "d26df342-1ef0-4776-9e33-12688ea2b283"
}
```
```
2018-11-15T18:07:26.404Z	4edd1195-e901-11e8-8e36-7f76855d6589	process_job_event:
{
    "rc": 0,
    "resource": "job",
    "event": "output",
    "value": {
        "input": {
            "graph": {},
            "Rosenbrock": {
                "x2": 4.7414715965461,
                "x3": 1.9728632790287541,
                "x1": 2.147071494550927,
                "x6": -4.273527432151759,
                "x4": -1.55748612478315,
                "x5": -0.6412897031329639
            }
        },
        "nodeError": {
            "Rosenbrock": 0
        },
        "nodeSettings": {
            "Rosenbrock": {}
        },
        "output": {
            "graph": {
                "error": 0
            },
            "Rosenbrock": {
                "f": 48193.15509334679
            }
        },
        "turbineMessages": {
            "Rosenbrock": "NULL"
        },
        "graphError": 0,
        "solTime": 0
    },
    "jobid": "d26df342-1ef0-4776-9e33-12688ea2b283"
}
```

**Comment by boverhof** [comment]
## FIX
In the results generator check to see that if "item.success" is not undefined also check "item.output" is not undefined.

```
foqus-boverhof % git diff cloud/aws/node/post-session-result/index.js
diff --git a/cloud/aws/node/post-session-result/index.js b/cloud/aws/node/post-session-result/index.js
index d00a8f7..2d8158f 100644
--- a/cloud/aws/node/post-session-result/index.js
+++ b/cloud/aws/node/post-session-result/index.js
@@ -121,7 +121,7 @@ exports.handler = function(event, context, callback) {
             console.log('Data: ', data.Items.length);
             for (var i=0; i<data.Items.length; i++) {
                 var item = data.Items[i];
-                if (item.success != undefined) {
+                if (item.success != undefined && item.output != undefined) {
                   console.log('success item: ', item);
                   finished_jobs[item.Id] = item;
                 } else if (item.error != undefined) {
foqus-boverhof %
```

**Comment by boverhof** [comment]
##  Rosenbrock example
### Failure
changed the solver option to "OptCMA", the optimization failed during the 3rd iteration (around 15 runs )
### Confirmed Fix
Added the check to "item.output" above and the optimization run to 130 iterations ( around 700 runs ).

**Comment by ksbeattie** [comment]
@boverhof says this is fixed and will update this with the commit that fixed it.

**Comment by boverhof** [comment]
This is for @ksbeattie 

https://github.com/CCSI-Toolset/FOQUS/pull/329

**Comment by ksbeattie** [comment]
Closing this on @boverhof 's word that this is fixed with PR #329
